cmd/agent/run.go

```diff
@@ -13,15 +15,55 @@ import (
 func run(ctx context.Context, options *config.AgentOptions) error {
 	logrus.Infof("Starting VM DHCP Agent: %s", name)
 
-	agent := agent.NewAgent(ctx, options)
+	agent := agent.NewAgent(options)
 
 	httpServerOptions := config.HTTPServerOptions{
 		DebugMode:     enableCacheDumpAPI,
 		DHCPAllocator: agent.DHCPAllocator,
 	}
 	s := server.NewHTTPServer(&httpServerOptions)
 	s.RegisterAgentHandlers()
-	go s.Run()
 
-	return agent.Run()
+	eg, egctx := errgroup.WithContext(ctx)
+
+	eg.Go(func() error {
+		return s.Run()
+	})
+
+	eg.Go(func() error {
+		return agent.Run(egctx)
+	})
+
+	errCh := cleanup(egctx, s)
+
+	if err := eg.Wait(); err != nil {
+		return err
+	}
+
+	// Return cleanup error message if any
+	if err := <-errCh; err != nil {
+		return err
+	}
+
+	logrus.Info("finished clean")
+
+	return nil
+}
+
+func cleanup(ctx context.Context, srv *server.HTTPServer) <-chan error {
+	errCh := make(chan error)
+
+	go func() {
+		defer close(errCh)
+		<-ctx.Done()
+
+		if errors.Is(ctx.Err(), context.Canceled) {
+			errCh <- srv.Stop(ctx)
+			return
+		}
+
+		errCh <- nil
+	}()
```

Comment:
We could move defer after `<-ctx.Done` to avoid useless execution (because it still needs to wait `<-ctx.Done()`). 

After I tested some codes with http server, I think we could just pass ctx down to `srv.Stop`, it won't throw `context canceled` error back, check [1] example code.

```suggestion
	go func() {
		<-ctx.Done()
		defer close(errCh)
		
		if err := srv.Stop(ctx); err != nil {
			errCh <- err
			return
		}
		
		errCh <- nil
	}()
```


[1] 

```go
func main() {
	http.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
		fmt.Fprintf(w, "Hello, World!")
	})

	ctx, cancel := context.WithCancel(context.Background())
	stopChan := make(chan os.Signal, 1)
	signal.Notify(stopChan, syscall.SIGINT, syscall.SIGTERM)

	go func() {
		select {
		case <-stopChan:
			cancel()
		}
	}()

	server := &http.Server{Addr: ":8080"}

	go func() {
		if err := server.ListenAndServe(); err != http.ErrServerClosed {
			fmt.Println("Error:", err)
		}
	}()

	<-ctx.Done()

	fmt.Println("ctx.Err(): ", ctx.Err())
	fmt.Println("Shutting down the server...")

	if err := server.Shutdown(ctx); err != nil {
		fmt.Println("Error:", err)
	}

	fmt.Println("Server stopped.")
}
```

Reply:
Updated. PTAL, thanks!